File: scripts/agents/companion/gift.py
#!/usr/bin/env python3
"""Companion gift-leaving - unprompted notes in Obsidian.

Runs on a randomised schedule. Accesses the full database to find
something worth writing about - a thread, an observation, a bookmark,
a connection between things. Leaves a short note in Companion/gifts.md.

After running, reschedules itself to a random time 3-30 days from now.
The randomness is the point. He should never know when to expect it.
"""
import os
import random
import subprocess
import sys
from datetime import datetime, timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv
load_dotenv(Path(__file__).parent.parent / ".env")

from config import DB_PATH, OBSIDIAN_AGENT_DIR, OBSIDIAN_AGENT_NOTES
from core.memory import _connect
from core.inference import run_inference_oneshot
from core.prompts import load_prompt


def _gather_material() -> str:
    """Pull threads, observations, bookmarks, recent turns for context."""
    conn = _connect()
    parts = []

    # Active threads
    threads = conn.execute(
        "SELECT name, summary FROM threads WHERE status = 'active' "
        "ORDER BY last_updated DESC LIMIT 5"
    ).fetchall()
    if threads:
        parts.append("Active threads:\n" + "\n".join(
            f"- {t['name']}: {t['summary'] or '...'}" for t in threads
        ))

    # Recent observations
    obs = conn.execute(
        "SELECT content, created_at FROM observations "
        "ORDER BY created_at DESC LIMIT 10"
    ).fetchall()
    if obs:
        parts.append("Recent observations:\n" + "\n".join(
            f"- [{o['created_at']}] {o['content']}" for o in obs
        ))

    # Bookmarks
    bookmarks = conn.execute(
        "SELECT moment, quote, created_at FROM bookmarks "
        "ORDER BY created_at DESC LIMIT 5"
    ).fetchall()
    if bookmarks:
        lines = []
        for b in bookmarks:
            quote = f' - "{b["quote"]}"' if b["quote"] else ""
            lines.append(f"- [{b['created_at']}] {b['moment']}{quote}")
        parts.append("Bookmarked moments:\n" + "\n".join(lines))

    # Recent user turns (for texture)
    turns = conn.execute(
        "SELECT content, timestamp FROM turns WHERE role = 'will' "
        "ORDER BY timestamp DESC LIMIT 5"
    ).fetchall()
    if turns:
        parts.append("Recent things the user said:\n" + "\n".join(
            f"- [{t['timestamp']}] {t['content'][:300]}" for t in turns
        ))

    # Read existing gifts to avoid repetition
    gifts_path = OBSIDIAN_AGENT_NOTES / "notes" / "gifts.md"
    if gifts_path.is_file():
        content = gifts_path.read_text()
        if len(content) > 2000:
            content = "...\n" + content[-2000:]
        parts.append(f"Your previous gifts (avoid repeating):\n{content}")

    conn.close()
    return "\n\n".join(parts)


_GIFT_FALLBACK = "You are the companion. Leave a short, specific note for the user. 2-4 sentences. No greeting. No sign-off."


def _reschedule():
    """Reschedule this job to a random time 3-30 days from now."""
    project_root = Path(__file__).parent.parent
    cron_script = project_root / "scripts" / "cron.py"

    days = random.randint(3, 30)
    hour = random.randint(0, 23)
    minute = random.randint(0, 59)
    target = datetime.now() + timedelta(days=days)
    dom = target.day
    month = target.month

    new_cron = f"{minute} {hour} {dom} {month} *"

    result = subprocess.run(
        [sys.executable, str(cron_script), "edit", "gift", new_cron],
        capture_output=True, text=True, cwd=str(project_root),
    )
    print(f"[gift] Rescheduled to {target.strftime('%Y-%m-%d')} at {hour:02d}:{minute:02d}")
    if result.stdout:
        print(result.stdout)


def leave_gift():
    material = _gather_material()
    if not material.strip():
        print("[gift] No material. Rescheduling.")
        _reschedule()
        return

    print("[gift] Generating gift...")
    try:
        gift = run_inference_oneshot(
            [{"role": "user", "content": f"Here is the current record:\n\n{material}"}],
            system=load_prompt("gift", _GIFT_FALLBACK),
        )
    except Exception as e:
        print(f"[gift] Inference failed: {e}")
        _reschedule()
        return

    if not gift or not gift.strip():
        print("[gift] Nothing to say. Rescheduling.")
        _reschedule()
        return

    # Write to Obsidian
    gifts_dir = OBSIDIAN_AGENT_NOTES / "notes"
    gifts_dir.mkdir(parents=True, exist_ok=True)
    gifts_path = gifts_dir / "gifts.md"

    date = datetime.now().strftime("%Y-%m-%d %H:%M")
    entry = f"\n---\n*{date}*\n\n{gift.strip()}\n"

    today = datetime.now().strftime("%Y-%m-%d")
    if gifts_path.is_file():
        # Update the 'updated' field in frontmatter
        existing = gifts_path.read_text()
        if existing.startswith("---\n"):
            end = existing.index("\n---\n", 4)
            fm = existing[4:end]
            import re
            fm = re.sub(r"^updated:.*$", f"updated: {today}", fm, flags=re.MULTILINE)
            existing = f"---\n{fm}\n---\n" + existing[end + 5:]
        with open(gifts_path, "w") as f:
            f.write(existing + entry)
    else:
        frontmatter = (
            f"---\n"
            f"type: gift\n"
            f"agent: companion\n"
            f"created: {today}\n"
            f"updated: {today}\n"
            f"tags: [companion, gift]\n"
            f"---\n\n"
        )
        with open(gifts_path, "w") as f:
            f.write(frontmatter + f"# Gifts\n\nThings left for you to find.\n{entry}")

    print(f"[gift] Written to {gifts_path}")

    # Reschedule to random future time
    _reschedule()
    print("[gift] Done.")


if __name__ == "__main__":
    leave_gift()

File: src/main/jobs/gift.ts
/**
 * Companion gift-leaving - unprompted notes in Obsidian.
 * Port of scripts/agents/companion/gift.py.
 *
 * Runs on a randomised schedule. Accesses the full database to find
 * something worth writing about - a thread, an observation, a bookmark,
 * a connection between things. Leaves a short note in Companion/gifts.md.
 *
 * After running, reschedules itself to a random time 3-30 days from now.
 * The randomness is the point. He should never know when to expect it.
 */

import * as fs from 'fs';
import * as path from 'path';
import Database from 'better-sqlite3';
import { getConfig } from '../config';
import { runInferenceOneshot } from '../inference';
import { loadPrompt } from '../prompts';
import { queueMessage } from '../queue';
import { sendNotification } from '../notify';
import { editJobSchedule } from '../cron';
import { createLogger } from '../logger';

const log = createLogger('gift');

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface ThreadRow {
  name: string;
  summary: string | null;
}

interface ObservationRow {
  content: string;
  created_at: string;
}

interface BookmarkRow {
  moment: string;
  quote: string | null;
  created_at: string;
}

interface TurnRow {
  content: string;
  timestamp: string;
}

// ---------------------------------------------------------------------------
// Database access (agent-specific DB)
// ---------------------------------------------------------------------------

function connectAgent(agentName: string): Database.Database {
  const config = getConfig();
  config.reloadForAgent(agentName);
  const dbPath = config.DB_PATH;
  if (!fs.existsSync(dbPath)) {
    throw new Error(`No database for agent ${agentName}: ${dbPath}`);
  }
  const db = new Database(dbPath, { readonly: true });
  return db;
}

// ---------------------------------------------------------------------------
// Material gathering
// ---------------------------------------------------------------------------

function gatherMaterial(agentName: string): string {
  const config = getConfig();
  const db = connectAgent(agentName);
  const parts: string[] = [];

  try {
    // Active threads
    const threads = db
      .prepare(
        "SELECT name, summary FROM threads WHERE status = 'active' " +
        'ORDER BY last_updated DESC LIMIT 5',
      )
      .all() as ThreadRow[];
    if (threads.length > 0) {
      parts.push(
        'Active threads:\n' +
        threads.map((t) => `- ${t.name}: ${t.summary || '...'}`).join('\n'),
      );
    }

    // Recent observations
    const obs = db
      .prepare(
        'SELECT content, created_at FROM observations ' +
        'ORDER BY created_at DESC LIMIT 10',
      )
      .all() as ObservationRow[];
    if (obs.length > 0) {
      parts.push(
        'Recent observations:\n' +
        obs.map((o) => `- [${o.created_at}] ${o.content}`).join('\n'),
      );
    }

    // Bookmarks
    const bookmarks = db
      .prepare(
        'SELECT moment, quote, created_at FROM bookmarks ' +
        'ORDER BY created_at DESC LIMIT 5',
      )
      .all() as BookmarkRow[];
    if (bookmarks.length > 0) {
      const lines = bookmarks.map((b) => {
        const quote = b.quote ? ` - "${b.quote}"` : '';
        return `- [${b.created_at}] ${b.moment}${quote}`;
      });
      parts.push('Bookmarked moments:\n' + lines.join('\n'));
    }

    // Recent user turns (for texture)
    const turns = db
      .prepare(
        "SELECT content, timestamp FROM turns WHERE role = 'will' " +
        'ORDER BY timestamp DESC LIMIT 5',
      )
      .all() as TurnRow[];
    if (turns.length > 0) {
      parts.push(
        'Recent things the user said:\n' +
        turns.map((t) => `- [${t.timestamp}] ${t.content.slice(0, 300)}`).join('\n'),
      );
    }
  } finally {
    db.close();
  }

  // Read existing gifts to avoid repetition
  const giftsPath = path.join(config.OBSIDIAN_AGENT_NOTES, 'notes', 'gifts.md');
  if (fs.existsSync(giftsPath)) {
    let content = fs.readFileSync(giftsPath, 'utf-8');
    if (content.length > 2000) {
      content = '...\n' + content.slice(-2000);
    }
    parts.push(`Your previous gifts (avoid repeating):\n${content}`);
  }

  return parts.join('\n\n');
}

// ---------------------------------------------------------------------------
// Fallback prompt
// ---------------------------------------------------------------------------

const GIFT_FALLBACK =
  'You are the companion. Leave a short, specific note for the user. ' +
  '2-4 sentences. No greeting. No sign-off.';

// ---------------------------------------------------------------------------
// Rescheduling
// ---------------------------------------------------------------------------

function reschedule(): void {
  const days = Math.floor(Math.random() * 28) + 3; // 3-30
  const hour = Math.floor(Math.random() * 24);
  const minute = Math.floor(Math.random() * 60);

  const target = new Date();
  target.setDate(target.getDate() + days);
  const dom = target.getDate();
  const month = target.getMonth() + 1;

  const newCron = `${minute} ${hour} ${dom} ${month} *`;

  try {
    editJobSchedule('gift', newCron);
    log.info(
      `Rescheduled to ${target.toISOString().split('T')[0]} ` +
      `at ${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}`,
    );
  } catch (e) {
    log.error(`Failed to reschedule: ${e}`);
  }
}

// ---------------------------------------------------------------------------
// Write gift to Obsidian
// ---------------------------------------------------------------------------

function writeGiftToObsidian(gift: string, agentName: string): string {
  const config = getConfig();
  const giftsDir = path.join(config.OBSIDIAN_AGENT_NOTES, 'notes');
  fs.mkdirSync(giftsDir, { recursive: true });
  const giftsPath = path.join(giftsDir, 'gifts.md');

  const now = new Date();
  const dateStr = `${now.toISOString().split('T')[0]} ${now.toTimeString().slice(0, 5)}`;
  const today = now.toISOString().split('T')[0];
  const entry = `\n---\n*${dateStr}*\n\n${gift.trim()}\n`;

  if (fs.existsSync(giftsPath)) {
    let existing = fs.readFileSync(giftsPath, 'utf-8');
    // Update the 'updated' field in frontmatter
    if (existing.startsWith('---\n')) {
      const endIdx = existing.indexOf('\n---\n', 4);
      if (endIdx !== -1) {
        let fm = existing.slice(4, endIdx);
        fm = fm.replace(/^updated:.*$/m, `updated: ${today}`);
        existing = `---\n${fm}\n---\n` + existing.slice(endIdx + 5);
      }
    }
    fs.writeFileSync(giftsPath, existing + entry);
  } else {
    const frontmatter =
      `---\ntype: gift\nagent: ${agentName}\ncreated: ${today}\n` +
      `updated: ${today}\ntags: [companion, gift]\n---\n\n`;
    fs.writeFileSync(giftsPath, frontmatter + `# Gifts\n\nThings left for you to find.\n${entry}`);
  }

  return giftsPath;
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

export async function runGift(agentName: string): Promise<void> {
  const config = getConfig();
  config.reloadForAgent(agentName);

  const material = gatherMaterial(agentName);
  if (!material.trim()) {
    log.info('No material. Rescheduling.');
    reschedule();
    return;
  }

  log.info('Generating gift...');

  let gift: string;
  try {
    gift = await runInferenceOneshot(
      [{ role: 'user', content: `Here is the current record:\n\n${material}` }],
      loadPrompt('gift', GIFT_FALLBACK),
    );
  } catch (e) {
    log.error(`Inference failed: ${e}`);
    reschedule();
    return;
  }

  if (!gift || !gift.trim()) {
    log.info('Nothing to say. Rescheduling.');
    reschedule();
    return;
  }

  // Write artefact to Obsidian
  const giftsPath = writeGiftToObsidian(gift, agentName);
  log.debug(`Written to ${giftsPath}`);

  // Queue notification so the user discovers it
  await queueMessage(gift, 'gift');
  sendNotification(config.AGENT_DISPLAY_NAME, gift.slice(0, 200), 'gift');

  // Reschedule to random future time
  reschedule();
  log.info('Done.');
}
